### PR TITLE
(feat) Enhanced orientation breadcrumbs with explicit "from" and "to"

### PR DESCRIPTION
**Pull request by robinmacharg**
## Goal

Report both the old and new orientation values as metadata. Match the casing for the breadcrumb message.

## Design

On device orientation change notification, record the current device orientation to be used the next time orientation changes.  When both current and previous orientations are non-nil and differ record an appropriate breadcrumb.

## Changeset

The change is located in `BugsnagClient`.

## Tests

This is difficult to test.  For Unit testing mocking would allow the UIDevice to be stubbed out and simulate a rotation.  For CI/e2e tests the possibility exists (but has not been proven to work) of sending a rotation keypress (Cmd-Left) to the simulator.

The implementation is simple and in the interests of time this feature is left as a manual test until mocking libraries have been added.

### Outstanding Questions

- This pull request is ready for:
  - [ ] Initial review of the intended approach, not yet feature complete
  - [ ] Structural review of the classes, functions, and properties modified
  - [X] Final review

- [ ] The correct target branch has been selected (`master` for fixes, `next` for
  features)
- [ ] Consistency across platforms for structures or concepts added or modified
- [ ] Consistency between the changeset and the goal stated above
- [ ] Internal consistency with the rest of the library - is there any overlap between existing interfaces and any which have been added?
- [ ] Usage friction - is the proposed change in usage cumbersome or complicated?
- [ ] Performance and complexity - are there any cases of unexpected O(n^3) when iterating, recursing, flat mapping, etc?
- [X] Concurrency concerns - if components are accessed asynchronously, what issues will arise
- [ ] Thoroughness of added tests and any missing edge cases
- [ ] Idiomatic use of the language

### CODE REVIEW

File: Source/BugsnagClient.m [hunk 1]
```diff
@@ -811,14 +815,26 @@ - (void)orientationChanged:(NSNotification *)notif {
...
+        if ([orientation isEqualToString:metadata[BSGKeyOrientation]])
             return; // ignore duplicate orientation event
-        }
     }
 
-    [[self state] addAttribute:BSGKeyOrientation
-                     withValue:orientation
-                 toTabWithName:BSGKeyDeviceState];
+    // It's not a change
+    if ([orientation isEqualToString:lastOrientation])
```
Comment (fractalwrench) [inline Source/BugsnagClient.m]: nit: could braces be added to new if statements throughout this file?

I don't know if we actually have any official code style guide for Cocoa, so maybe that's something to discuss.

File: Source/BugsnagClient.m [hunk 2]
```diff
@@ -84,6 +84,10 @@
 static NSUInteger handledCount;
 static NSUInteger unhandledCount;
 static bool hasRecordedSessions;
+#if TARGET_IPHONE_SIMULATOR || TARGET_OS_IPHONE
+// The previous device orientation - iOS only
+static NSString *lastOrientation = NULL;
```
Comment (fractalwrench) [inline Source/BugsnagClient.m]: Can this be declared as a property instead? This would make it easier to get to a point where it'd be possible to instantiate multiple `BugsnagClient` objects, and also simplify any future extraction of this logic to a separate class.

File: Source/BugsnagClient.m [hunk 3]
```diff
@@ -811,14 +815,26 @@ - (void)orientationChanged:(NSNotification *)notif {
...
 
-    [[self state] addAttribute:BSGKeyOrientation
-                     withValue:orientation
-                 toTabWithName:BSGKeyDeviceState];
+    // It's not a change
+    if ([orientation isEqualToString:lastOrientation])
+        return;
+
+    // We previously had an orientation
+    if (lastOrientation) {
```
Comment (fractalwrench) [inline Source/BugsnagClient.m]: This can be simplified by calculating the `lastOrientation` when `BugsnagClient` is initialised (assuming that the return value of `[UIDevice currentDevice].orientation` is non-null)

File: Source/BugsnagClient.m (hunk 1 above)
Comment (robinmacharg) [reply Source/BugsnagClient.m]: We don't.  I've been taking the arbitrary decision that single line conditionals don\t need the baggage of a closing brace.  I can see the other side of the argument.  I'll leave as is and raise a ticket to come up with a definitive style guide.

Comment (snmaynard) [reply Source/BugsnagClient.m]: The style guide for C code is pretty universally always include braces around if statements because it's easy to miss that they are missing when adding future code.

File: Source/BugsnagClient.m (hunk 3 above)
Comment (robinmacharg) [reply Source/BugsnagClient.m]: We can't assume this.  Acording to [the docs](https://developer.apple.com/documentation/uikit/uidevice/1620041-begingeneratingdeviceorientation?language=objc) `beginGeneratingDeviceOrientationNotifications()` needs called.  We do this in `BugsnagClient.start()`.  As the code stands we call `BugsnagClient.initWithConfiguration()` first, then `BugsnagClient.start()`.  We could take the golden path and short-circuit on the first call instead.

File: Source/BugsnagClient.m (hunk 2 above)
Comment (robinmacharg) [reply Source/BugsnagClient.m]: I'm assuming you mean class property?  Yes, that makes sense.

Comment (fractalwrench) [reply Source/BugsnagClient.m]: Yep, that's correct

File: Source/BugsnagClient.m (hunk 3 above)
Comment (fractalwrench) [reply Source/BugsnagClient.m]: Isn't it possible to move where `beginGeneratingDeviceOrientationNotifications()` is called to make calculation of an initial value feasible?

File: Source/BugsnagClient.m (hunk 1 above)
Comment (robinmacharg) [reply Source/BugsnagClient.m]: Task created to agree on Cocoa style.  Braces will be readded.

File: Source/BugsnagClient.m (hunk 3 above)
Comment (robinmacharg) [reply Source/BugsnagClient.m]: Yes, but I don't believe that there's any guarantee that the orientation has been determined at any specific time.  A test implementation bears this out.  One way round this deficit would be to use `[[UIApplication sharedApplication] statusBarOrientation]` but this does not allow for face up/down determination.  If the goal is to have a breadcrumb with valid from/to values I think the current approach with an early exit is the most consistent.

Comment (fractalwrench) [reply Source/BugsnagClient.m]: Happy enough to go with the current implementation if it's not always feasible to get an initial value.

File: Source/BugsnagClient.m [hunk 4]
```diff
@@ -774,51 +829,46 @@ - (void)batteryChanged:(NSNotification *)notif {
...
         }
     }
-
-    [[self state] addAttribute:BSGKeyOrientation
-                     withValue:orientation
+    
+    // Record the breadcrumb
+    [[self state] addAttribute:BSGKeyOrientationChange
+                     withValue:@{@"from" : _lastOrientation,
+                                 @"to" : orientation}
```
Comment (tomlongridge) [inline Source/BugsnagClient.m]: This is the call to add metadata - think won't be adding it to the breadcrumb (as far as I can see), this seems to be done through `sendBreadcrumbForNotification`.

File: Source/BugsnagKeys.h [hunk 5]
```diff
@@ -43,6 +43,7 @@ static NSString *const BSGKeyLabel = @"label";
 static NSString *const BSGKeySeverityReason = @"severityReason";
 static NSString *const BSGKeyLogLevel = @"logLevel";
 static NSString *const BSGKeyOrientation = @"orientation";
+static NSString *const BSGKeyOrientationChange = @"Orientation change";
```
Comment (tomlongridge) [inline Source/BugsnagKeys.h]: I'm not sure this is really a key - I think it should use the notification name.

File: Source/BugsnagClient.m [hunk 6]
```diff
@@ -42,12 +42,6 @@
 #import "BSG_KSSystemInfo.h"
 #import "BSG_KSMach.h"
 
-#if TARGET_IPHONE_SIMULATOR || TARGET_OS_IPHONE
```
Comment (fractalwrench) [inline Source/BugsnagClient.m]: Just wanted to check why it's necessary to move these imports to the header file?

File: Source/BugsnagClient.m [hunk 7]
```diff
@@ -217,6 +252,23 @@ @interface BugsnagConfiguration ()
 
 @implementation BugsnagClient
 
+#if TARGET_IPHONE_SIMULATOR || TARGET_OS_IPHONE
+/**
+ * Storage for the device orientation.  It is "last" whenever an orientation change is received
+ */
+static NSString *_lastOrientation = nil;
```
Comment (fractalwrench) [inline Source/BugsnagClient.m]: Why does this need to be static? Is it possible to just assign to `self.lastOrientation` declared in the class extension instead?

File: Source/BugsnagClient.m [hunk 8]
```diff
@@ -774,51 +829,46 @@ - (void)batteryChanged:(NSNotification *)notif {
...
+    else if (!_lastOrientation || [orientation isEqualToString:_lastOrientation]) {
+        _lastOrientation = orientation;
+        return;
     }
 
-    NSDictionary *lastBreadcrumb =
-        [[self.configuration.breadcrumbs arrayValue] lastObject];
-    NSString *orientationNotifName =
-        BSGBreadcrumbNameForNotificationName(notif.name);
+    NSDictionary *lastBreadcrumb = [[self.configuration.breadcrumbs arrayValue] lastObject];
```
Comment (fractalwrench) [inline Source/BugsnagClient.m]: This feels error-prone - users could choose to reject the breadcrumb via an `OnBreadcrumb` callback. It should also be possible to just use `self.lastOrientation` to remove duplicate orientation events, unless I've missed some additional complexity here.

File: Source/BugsnagKeys.h (hunk 5 above)
Comment (robinmacharg) [reply Source/BugsnagKeys.h]: It's not a key per-se, but it is a constant string.  `keys.h` was an unobtrusive place to put it.  As per other constant strings scattered throughout the library I'll put it in `client.m`.

File: Source/BugsnagClient.m (hunk 6 above)
Comment (robinmacharg) [reply Source/BugsnagClient.m]: Good spot.  Hangover from trying to make `orientationChanged` testable.  Moved back.

File: Source/BugsnagClient.m (hunk 7 above)
Comment (robinmacharg) [reply Source/BugsnagClient.m]: Previous discussion agreed that this should be a class property.  My understanding is that class properties need static storage.  Am I misunderstanding something?

Comment (fractalwrench) [reply Source/BugsnagClient.m]: Apologies - my terminology might be a little off as it's a while since I've used Objective-C in anger. The intent behind my review comment is that this would be defined in a [class extension](https://developer.apple.com/library/archive/documentation/Cocoa/Conceptual/ProgrammingWithObjectiveC/CustomizingExistingClasses/CustomizingExistingClasses.html#//apple_ref/doc/uid/TP40011210-CH6-SW3), similar to this:

```
// within BugsnagClient.m

@interface BugsnagClient()
@property NSString *lastOrientation;
@end
```

This would avoid exposing the property publicly. Adding new static variables will make it harder to instantiate multiple objects of `BugsnagClient` - which is something we may want to do eventually. As far as I'm aware using a class extension wouldn't require static storage?

File: Source/BugsnagClient.m (hunk 4 above)
Comment (robinmacharg) [reply Source/BugsnagClient.m]: Doh.  Resolved after discussion.

File: Source/BugsnagKeys.h (hunk 5 above)
Comment (robinmacharg) [reply Source/BugsnagKeys.h]: Resolved after discussion.  Removed, replaced by a key in `notificationNameMap`.

File: Source/BugsnagClient.m (hunk 8 above)
Comment (robinmacharg) [reply Source/BugsnagClient.m]: Restructured the entire method after discussion.

File: Source/BugsnagClient.m (hunk 7 above)
Comment (robinmacharg) [reply Source/BugsnagClient.m]: Discussed offline, moved from class to instance-level storage.
